{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "74c878b5-cf48-42cc-8db7-19371adc4276",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c9f9cd72-5a0f-4acf-841d-e20d1979ba96",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from hmmlearn import hmm\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "from numpy.linalg import matrix_power\n",
    "from collections import Counter\n",
    "from scipy.stats import poisson\n",
    "from scipy.stats import linregress\n",
    "\n",
    "#https://archive.ics.uci.edu/ml/datasets/Molecular+Biology+%28Promoter+Gene+Sequences%29\n",
    "data = Path('p.datas')\n",
    "promoters = open(data, \"r\")\n",
    "d = promoters.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c7bb174a-9e06-4f02-b8dc-abde67d894c6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open('p.data', 'r') as f:\n",
    "    # Read the file line by line\n",
    "    data1 = f.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "49c54a84-c117-4c33-9861-6344467bbc3a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequences = []\n",
    "# Iterate over the lines and split each line using a comma delimiter\n",
    "for line in data1:\n",
    "    # Remove any trailing newline characters\n",
    "    line = line.rstrip('\\n')\n",
    "    line = line.rstrip('\\t')\n",
    "\n",
    "    # Split the line into a list of values using a comma delimiter\n",
    "    values = line.split(',')\n",
    "    val = values[2].strip('\\t').upper()\n",
    "    sequences.append(val)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2ceb74a-e233-4dd4-84a0-1dc30f92f433",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Markov Chains"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948e8350-c11f-4dd6-9f4a-ffbb4a51c27c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ddc85020-6659-47af-981f-fce4c4646da0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "nucleotides = ['A', 'G', 'C', 'T']\n",
    "total_permuts = []\n",
    "\n",
    "# Determines total permutations of nucleotides.\n",
    "# Ex. AAAA, AGCT, CGTA, TCGA, etc.\n",
    "def permut(nucleotides, curr_permut, length):\n",
    "    if length == 0:\n",
    "        total_permuts.append(curr_permut)\n",
    "    else:\n",
    "        for n in nucleotides:\n",
    "            permut(nucleotides, curr_permut + n, length - 1)\n",
    "            \n",
    "# Needed for computing probabilties\n",
    "permut(nucleotides, '', 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "28b61369-6e38-4802-ba7b-2407a003799d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# This function will calculate the probabilities of the probabilty matrix\n",
    "def compute_probs(arr_seqs):\n",
    "    # Define a 4x4 matrix to store the transition counts\n",
    "    transition_counts = np.zeros((4, 4))\n",
    "\n",
    "    # Iterate over all permutations of nucleotides the dataset and count transitions\n",
    "    for a in total_permuts:\n",
    "        # Iterate over all sequecnes from the dataset\n",
    "        for seq in arr_seqs:\n",
    "            curr_nucleotide = seq[0]\n",
    "            next_nucleotide = seq[1]\n",
    "            for i in range(len(seq)-1):\n",
    "                # Count Transitions\n",
    "                if curr_nucleotide not in nucleotides:\n",
    "                    curr_nucleotide = seq[i]\n",
    "                    continue\n",
    "                if next_nucleotide not in nucleotides:\n",
    "                    curr_nucleotide = seq[i]\n",
    "                    next_nucleotide = seq[i+1]\n",
    "                    continue\n",
    "                curr_count = nucleotides.index(curr_nucleotide)\n",
    "                next_count = nucleotides.index(next_nucleotide)\n",
    "                transition_counts[curr_count][next_count] += 1\n",
    "                curr_nucleotide = seq[i+1]\n",
    "                if i < len(seq)-2:\n",
    "                    next_nucleotide = seq[i+2]\n",
    "                    \n",
    "    # Compute the transition probabilities from the counts\n",
    "    transition_probs = transition_counts / np.sum(transition_counts, axis=1, keepdims=True)\n",
    "    return transition_probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "687e891e-5507-481f-ad5d-ad915466beb0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transition probabilities:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>0.281008</td>\n",
       "      <td>0.204134</td>\n",
       "      <td>0.243540</td>\n",
       "      <td>0.271318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>0.282414</td>\n",
       "      <td>0.186289</td>\n",
       "      <td>0.270492</td>\n",
       "      <td>0.260805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>0.252193</td>\n",
       "      <td>0.260234</td>\n",
       "      <td>0.194444</td>\n",
       "      <td>0.293129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>0.232420</td>\n",
       "      <td>0.258045</td>\n",
       "      <td>0.209774</td>\n",
       "      <td>0.299762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             A         G         C         T\n",
       "0  A  0.281008  0.204134  0.243540  0.271318\n",
       "1  G  0.282414  0.186289  0.270492  0.260805\n",
       "2  C  0.252193  0.260234  0.194444  0.293129\n",
       "3  T  0.232420  0.258045  0.209774  0.299762"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Print the transition probabilities\n",
    "print(\"Transition probabilities:\")\n",
    "all_seqs_probs = compute_probs(sequences)\n",
    "def transpose_probs_to_df(transition_probs):\n",
    "    transpose_probs = transition_probs.transpose()\n",
    "    d = {\"\" : nucleotides, \"A\" : list(transpose_probs[0]), \"G\" : list(transpose_probs[1]), \n",
    "         \"C\" : list(transpose_probs[2]), \"T\" : list(transpose_probs[3])}\n",
    "    df = pd.DataFrame(d)\n",
    "    return df\n",
    "df_all_seqs = transpose_probs_to_df(all_seqs_probs)\n",
    "display(df_all_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ecec67f5-e681-4dcb-a3fb-b2ce3b076bf0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transition probabilities:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>0.281008</td>\n",
       "      <td>0.204134</td>\n",
       "      <td>0.243540</td>\n",
       "      <td>0.271318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>0.282414</td>\n",
       "      <td>0.186289</td>\n",
       "      <td>0.270492</td>\n",
       "      <td>0.260805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>0.252193</td>\n",
       "      <td>0.260234</td>\n",
       "      <td>0.194444</td>\n",
       "      <td>0.293129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>0.232420</td>\n",
       "      <td>0.258045</td>\n",
       "      <td>0.209774</td>\n",
       "      <td>0.299762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             A         G         C         T\n",
       "0  A  0.281008  0.204134  0.243540  0.271318\n",
       "1  G  0.282414  0.186289  0.270492  0.260805\n",
       "2  C  0.252193  0.260234  0.194444  0.293129\n",
       "3  T  0.232420  0.258045  0.209774  0.299762"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Print the transition probabilities\n",
    "print(\"Transition probabilities:\")\n",
    "all_seqs_probs = compute_probs(sequences)\n",
    "def transpose_probs_to_df(transition_probs):\n",
    "    transpose_probs = transition_probs.transpose()\n",
    "    d = {\"\" : nucleotides, \"A\" : list(transpose_probs[0]), \"G\" : list(transpose_probs[1]), \n",
    "         \"C\" : list(transpose_probs[2]), \"T\" : list(transpose_probs[3])}\n",
    "    df = pd.DataFrame(d)\n",
    "    return df\n",
    "df_all_seqs = transpose_probs_to_df(all_seqs_probs)\n",
    "display(df_all_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b8b75efc-6ea7-4f6f-8269-6412cee5928e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>0.337772</td>\n",
       "      <td>0.158596</td>\n",
       "      <td>0.215496</td>\n",
       "      <td>0.288136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>0.235195</td>\n",
       "      <td>0.174281</td>\n",
       "      <td>0.292724</td>\n",
       "      <td>0.297800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>0.251135</td>\n",
       "      <td>0.249622</td>\n",
       "      <td>0.225416</td>\n",
       "      <td>0.273828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>0.268539</td>\n",
       "      <td>0.221348</td>\n",
       "      <td>0.176404</td>\n",
       "      <td>0.333708</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             A         G         C         T\n",
       "0  A  0.337772  0.158596  0.215496  0.288136\n",
       "1  G  0.235195  0.174281  0.292724  0.297800\n",
       "2  C  0.251135  0.249622  0.225416  0.273828\n",
       "3  T  0.268539  0.221348  0.176404  0.333708"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Positive Model means it is a promoter sequence\n",
    "promoters = sequences[:53]\n",
    "promoter_probs = compute_probs(promoters)\n",
    "display_promoter_probs = transpose_probs_to_df(promoter_probs)\n",
    "display(display_promoter_probs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "af1a526a-ad37-4e52-b8f1-5660e5868ea3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>0.216066</td>\n",
       "      <td>0.256233</td>\n",
       "      <td>0.275623</td>\n",
       "      <td>0.252078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>0.319574</td>\n",
       "      <td>0.195739</td>\n",
       "      <td>0.252996</td>\n",
       "      <td>0.231691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>0.253182</td>\n",
       "      <td>0.270156</td>\n",
       "      <td>0.165488</td>\n",
       "      <td>0.311174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>0.191624</td>\n",
       "      <td>0.299492</td>\n",
       "      <td>0.247462</td>\n",
       "      <td>0.261421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             A         G         C         T\n",
       "0  A  0.216066  0.256233  0.275623  0.252078\n",
       "1  G  0.319574  0.195739  0.252996  0.231691\n",
       "2  C  0.253182  0.270156  0.165488  0.311174\n",
       "3  T  0.191624  0.299492  0.247462  0.261421"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Negative Model means it is not a promoter sequence\n",
    "neg_promoters = sequences[53:]\n",
    "neg_promoter_probs = compute_probs(neg_promoters)\n",
    "display_neg_promoter_probs = transpose_probs_to_df(neg_promoter_probs)\n",
    "display(display_neg_promoter_probs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "626a884e-9260-4ba1-8c00-56c81f44074d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Log-Odds Ratio with Respect to Model Building: Measure of How Well the Model Predicts the Data\n",
    "$S(x) = \\log \\left( \\frac{P(x|\\text{model}^+)}{P(x|\\text{model}^-)} \\right) = \\sum_{i=1}^{L} \\log \\left( \\frac{a^+_{(x_{i-1})(x_i)}}{a^-_{(x_{i-1})(x_i)}} \\right)$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d898db0a-0c2d-4e4c-9401-38002a06e9d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The log-odds ratio of the model is -0.145 which implies both models predict a normal genome.\n"
     ]
    }
   ],
   "source": [
    "# Log-odds model to predict the if the model is deviates from the standard model\n",
    "S = 0\n",
    "index1 = 0\n",
    "index2 = 0\n",
    "\n",
    "for i in promoter_probs:\n",
    "    while index1 < 4:\n",
    "        for j in neg_promoter_probs:\n",
    "            while index2 < 4:\n",
    "                S += np.log( promoter_probs[index1][index2]/neg_promoter_probs[index1][index2] )\n",
    "                index2 += 1\n",
    "        index1 += 1\n",
    "print(\"The log-odds ratio of the model is\", round(S, 3), \"which implies both models predict a normal genome.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6202160d-7b99-4752-ac4c-ded2dfde262e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Gambler's Ruin Problem in the Perspective of Gene Sequences\n",
    "#### Consider an equilibrium reaction between methylating proteins and RNA polmerase proteins such that RNA polymerase begins transcription on the promoter region of a DNA strand and methylating proteins stop transcription. Methylating proteins have one binding site, but if it binds to the promoter region, the rate of another protein methylating the promoter region is 1. RNA polymerase competes with the binding site of the promoter such that it has a 50% chance of knocking off a methylating protein and binding. Once there are 5 methylating proteins binded to the promoter region, no RNA polymerase can bind, but if there is only RNA polymerase binded to the promoter region, then no more methylating proteins will be able to bind. \n",
    "<br> 1. Describe this problem as a Markov chain using the methylating proteins. \n",
    "<br>2. What are the states? \n",
    "<br>3. Find the probability matrix."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ac9b204-e9c8-4938-a640-e6081124fb3d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Let $X_n$ be defined as a Markov Chain such that n represents the number of methylating proteins.\n",
    "#### $X_n$ = {1, 2, 3, 4, 5}\n",
    "#### The states are RNA polymerase binding (-1) or the methylating protein binding (+1)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "829a2988-19ec-462c-84d3-89111a826726",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1. , 0. , 0. , 0. , 0. , 0. ],\n",
       "       [0.5, 0. , 0.5, 0. , 0. , 0. ],\n",
       "       [0. , 0.5, 0. , 0.5, 0. , 0. ],\n",
       "       [0. , 0. , 0.5, 0. , 0.5, 0. ],\n",
       "       [0. , 0. , 0. , 0.5, 0. , 0.5],\n",
       "       [0. , 0. , 0. , 0. , 0. , 1. ]])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p00 = 1\n",
    "p01 = p02 = p03 = p04 = p05 = 0\n",
    "p10 = p12 = 0.5\n",
    "p11 = p13 = p14 = p15 = 0\n",
    "p21 = p23 = 0.5\n",
    "p20 = p22 = p24 = p25 = 0\n",
    "p32 = p34 = 0.5\n",
    "p30 = p31 = p33 = p35 = 0\n",
    "p43 = p45 = 0.5\n",
    "p40 = p41 = p42 = p44 = 0\n",
    "p50 = p51 = p52 = p53 = p54 = 0\n",
    "p55 = 1\n",
    "P = np.array([[p00, p01, p02, p03, p04, p05],\n",
    "                [p10, p11, p12, p13, p14, p15],\n",
    "                [p20, p21, p22, p23, p24, p25],\n",
    "                [p30, p31, p32, p33, p34, p35], \n",
    "                [p40, p41, p42, p43, p44, p45], \n",
    "                [p50, p51, p52, p53, p54, p55]\n",
    "             ])\n",
    "P"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ed74260-da7b-4d75-b956-6757dd428e4a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Using Chapman Kolmogrov Equation,\n",
    "### What is the probability that there are 4 methylating proteins consecutively attached to the promoter region after 10 events of competing methylating proteins and or RNA polymerases binding to the promoter region given the first protein to bind was a methylating protein?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "786e7d4b-970c-4a40-b237-a5cc56714099",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1.         0.         0.         0.         0.         0.        ]\n",
      " [0.75195312 0.03320312 0.         0.05371094 0.         0.16113281]\n",
      " [0.53710938 0.         0.08691406 0.         0.05371094 0.32226562]\n",
      " [0.32226562 0.05371094 0.         0.08691406 0.         0.53710938]\n",
      " [0.16113281 0.         0.05371094 0.         0.03320312 0.75195312]\n",
      " [0.         0.         0.         0.         0.         1.        ]]\n"
     ]
    }
   ],
   "source": [
    "from numpy.linalg import matrix_power\n",
    "P_10_steps = matrix_power(P, 10)\n",
    "print(P_10_steps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cf97a8bc-1e10-4b43-9fde-4fc8e439ff1e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P¹⁰₁₄ = 0.0537109375\n"
     ]
    }
   ],
   "source": [
    "print(f'P¹⁰₁₄ =', P_10_steps[1][3])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d84ce95-4bde-4ffe-af42-46d8d90c980b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Using a Markov Chain {W$_n$, n$\\in$$\\mathbb{N}$} with state space made up of states not in A and an extra state A."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f162dfc-283d-4556-96dc-db2e5071610b",
   "metadata": {},
   "source": [
    "#### Let N be the number of steps in a sequence until there is a GC pair. This is important to find in a DNA sequence as CpG islands have very high GC contents. If there is a presence of high GC contents near the promoter, then there is a CpG island which will promote methylation of the promoter region causing no transcription of the genes that are regulated by the promoter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15f0952-0c69-4ee2-a5cd-fb8e21e47ae1",
   "metadata": {},
   "source": [
    "There will be a repeat of GC pairs within 5 or less steps in the sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b9b8ba-a511-4078-b21a-5e430c3fc382",
   "metadata": {
    "tags": []
   },
   "source": [
    "P(N≤5) = P(W₅=C | W₀=G) = P$^5_{GC}$ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "df3f010e-fd8a-4b82-ad5d-23626bcdf794",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.26103354 0.22810559 0.22892887 0.281932  ]\n",
      " [0.26103362 0.2281053  0.2289292  0.28193189]\n",
      " [0.26103339 0.2281062  0.22892816 0.28193225]\n",
      " [0.26103344 0.22810599 0.2289284  0.28193217]]\n"
     ]
    }
   ],
   "source": [
    "seqs_5_steps = matrix_power(all_seqs_probs, 5)\n",
    "print(seqs_5_steps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a01a108a-2147-4623-86da-89206a095d65",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There is a 22.89 % chance that there will be a GC pair in less than 5 nucleotides.\n"
     ]
    }
   ],
   "source": [
    "print(\"There is a\", round(seqs_5_steps[1][2] * 100, 2), \"% chance that there will be a GC pair in less than 5 nucleotides.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "27412640-5539-405a-80c9-39825a3ab1a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.28100775, 0.20413437, 0.24354005, 0.27131783],\n",
       "       [0.28241431, 0.18628912, 0.2704918 , 0.26080477],\n",
       "       [0.25219298, 0.26023392, 0.19444444, 0.29312865],\n",
       "       [0.23241955, 0.25804529, 0.20977354, 0.29976162]])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_seqs_probs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "728f79df-1e74-46b9-83ce-dc578dc8b12f",
   "metadata": {},
   "source": [
    "## Long Run Probabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5c2e88-d1f3-4648-b984-6e8e6e91ecbf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "start = time.time()\n",
    "sum1 = 0\n",
    "sum2 = 0\n",
    "sum3 = 0\n",
    "sum4 = 0\n",
    "sums1_arr = []\n",
    "sums2_arr = []\n",
    "sums3_arr = []\n",
    "sums4_arr = []\n",
    "x = []\n",
    "slope_1 = []\n",
    "slope_2 = []\n",
    "slope_3 = []\n",
    "slope_4 = []\n",
    "\n",
    "for i in range(1, 2):\n",
    "    a = matrix_power(all_seqs_probs, i)[0][0]\n",
    "    b = matrix_power(all_seqs_probs, i)[1][1]\n",
    "    c = matrix_power(all_seqs_probs, i)[2][2]\n",
    "    d = matrix_power(all_seqs_probs, i)[3][3]\n",
    "    sum1 += a\n",
    "    sum2 += b\n",
    "    sum3 += c\n",
    "    sum4 += d\n",
    "    sums1_arr.append(sum1)\n",
    "    sums2_arr.append(sum2)\n",
    "    sums3_arr.append(sum3)\n",
    "    sums4_arr.append(sum4)\n",
    "    x.append(i)\n",
    "    slope_1.append(linregress(x, sums1_arr)[0])\n",
    "    slope_2.append(linregress(x, sums2_arr)[0])\n",
    "    slope_3.append(linregress(x, sums3_arr)[0])\n",
    "    slope_4.append(linregress(x, sums4_arr)[0])\n",
    "\n",
    "end = time.time()\n",
    "print(\"The time of execution of above program is :\",\n",
    "      (end-start) * 10**3, \"ms\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "118f1711-a7d0-4b49-9c8c-c0bfb7eaef67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a figure with two subplots (1 row, 2 columns)\n",
    "fig, (ax1, ax2, ax3, ax4) = plt.subplots(1, 2, figsize=(12, 6))\n",
    "\n",
    "# Plot on the first subplot\n",
    "ax1.plot(x, sums1_arr)\n",
    "ax1.set_title(\"Figure 1. Long Run $\\sum P_{ii}^n$ to $n=1,000,000$\")\n",
    "ax1.set_xlabel(\"n\")\n",
    "ax1.set_ylabel(\"$\\sum P_{ii}^n$\")\n",
    "\n",
    "# Plot on the second subplot\n",
    "ax2.plot(x, sums2_arr)\n",
    "ax2.set_title(\"Figure 2. Long Run $\\sum P_{ii}^n$ to $n=1,000,000$\")\n",
    "ax2.set_xlabel(\"n\")\n",
    "ax2.set_ylabel(\"$\\sum P_{ii}^n$\")\n",
    "\n",
    "ax3.plot(x, sums2_arr)\n",
    "ax3.set_title(\"Figure 3. Long Run $\\sum P_{ii}^n$ to $n=1,000,000$\")\n",
    "ax3.set_xlabel(\"n\")\n",
    "ax3.set_ylabel(\"$\\sum P_{ii}^n$\")\n",
    "\n",
    "ax4.plot(x, sums2_arr)\n",
    "ax4.set_title(\"Figure 4. Long Run $\\sum P_{ii}^n$ to $n=1,000,000$\")\n",
    "ax4.set_xlabel(\"n\")\n",
    "ax4.set_ylabel(\"$\\sum P_{ii}^n$\")\n",
    "\n",
    "# Adjust layout to prevent overlapping titles\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2dc82e8-6f28-4f62-ac36-dd903ec7d954",
   "metadata": {},
   "source": [
    "#### The slope for n=100,000 is [0.261, 0.228, 0.229, 0.282] which sums to 1.\n",
    "#### The slope for n=1,000,000 is [0.261, 0.228, 0.229, 0.282] which sums to 1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca8664e1-73cd-42b4-8d63-1ba62383a948",
   "metadata": {},
   "source": [
    "Below I have listed the long run proportions assuming the matrix is irreducible, aperiodic, ergodic, and stationary based on https://www2.ib.unicamp.br/profs/sfreis/SistematicaMolecular/Aula06SelecaoModelosSubstituicaoI/Leituras/The%20Phylogenetic%20Handbook%20-%20Chapter%204.pdf. The left column is calculated proportions from the slopes above and the right column is the calculated proportion using a calculator. \n",
    "\n",
    "Note: Chargaff's Rule states the %A≈%T and %C≈%T. The proportion of C and G is roughly equal which corresponds with Chargaff's Rule, but the proportion of A with respect to T is skewed to the %T.\n",
    "<br>$\\pi_A$: 0.261 |     0.261041\n",
    "<br>$\\pi_G$: 0.228 |     0.228112\n",
    "<br>$\\pi_C$: 0.229 |     0.228935\n",
    "<br>$\\pi_T$: 0.282 |     0.28194"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b67792b-6d3c-44b1-b7c1-a6506813d580",
   "metadata": {},
   "source": [
    "This is an approximation of the recurrent (ΣPᵢᵢⁿ = ∞) and transient (ΣPᵢᵢⁿ < ∞) definitions, and we find based on our assumptions that the ΣPᵢᵢⁿ is transient assuming our assumptions are true."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e888f9d5-18a3-4903-b448-affc6097f14d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Poisson Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ea99b81a-0448-4b67-8738-e47e8fa29271",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Used if you want to find poisson distribution based on the data\n",
    "def poisson_dist(mu, k):\n",
    "    return poisson.pmf(k, mu)\n",
    "\n",
    "counts = []\n",
    "for sequence in sequences:\n",
    "    counts.append(Counter(sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca0edbca-c657-4a3a-9fc5-5eb1cc98c1c7",
   "metadata": {},
   "source": [
    "We can use the maximum likelihood estimation (MLE) of the Poisson Distribution to model a Poisson Process. A MLE is a method that is used to find estimated parameters a distribution. Below is the MLE of the Poisson Distribution:\n",
    "\n",
    "Assume $X_1, X_2, ..., X_n$ are independent and identically distributed (i.i.d) random variables from a Poisson distribution with parameter $\\lambda$. \n",
    "\n",
    "The probability mass function for a Poisson distribution is given by:\n",
    "\n",
    "$$P(X = k) = \\frac{e^{-\\lambda}\\lambda^k}{k!}$$\n",
    "\n",
    "The likelihood function for $n$ independent observations is:\n",
    "\n",
    "\\begin{align*}\n",
    "L(\\lambda) &= P(X_1 = x_1, X_2 = x_2, ..., X_n = x_n | \\lambda)\\\\\n",
    "&= \\prod_{i=1}^{n} P(X_i = x_i | \\lambda)\\\\\n",
    "&= \\prod_{i=1}^{n} \\frac{e^{-\\lambda}\\lambda^{x_i}}{x_i!}\\\\\n",
    "&= \\frac{e^{-n\\lambda}\\lambda^{\\sum_{i=1}^{n}x_i}}{\\prod_{i=1}^{n}x_i!}\n",
    "\\end{align*}\n",
    "\n",
    "Then the natural logarithm of the likelihood function is taken:\n",
    "\n",
    "\\begin{align*}\n",
    "\\ln L(\\lambda) &= \\ln \\frac{e^{-n\\lambda}\\lambda^{\\sum_{i=1}^{n}x_i}}{\\prod_{i=1}^{n}x_i!}\\\\\n",
    "&= -n\\lambda + \\sum_{i=1}^{n}x_i \\ln \\lambda - \\sum_{i=1}^{n}\\ln(x_i!)\n",
    "\\end{align*}\n",
    "\n",
    "Differentiate with respect to $\\lambda$:\n",
    "\n",
    "\\begin{align*}\n",
    "\\frac{\\partial}{\\partial \\lambda} \\ln L(\\lambda) &= -n + \\frac{\\sum_{i=1}^{n}x_i}{\\lambda}\n",
    "\\end{align*}\n",
    "\n",
    "Set the derivative to zero to find the maximum likelihood estimate:\n",
    "\n",
    "\\begin{align*}\n",
    "\\frac{\\partial}{\\partial \\lambda} \\ln L(\\lambda) &= 0\\\\\n",
    "\\Rightarrow \\quad \\hat{\\lambda} &= \\frac{1}{n} \\sum_{i=1}^{n} x_i\n",
    "\\end{align*}\n",
    "\n",
    "The MLE of Poisson Distribution parameter $\\lambda$ has shown us that the sample mean is a good estimator of $\\lambda$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28022dd5-cd6e-46d5-a0b1-82d6885209cd",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Let $\\{N(t), t \\geq 0\\}$ be a Poisson Process such that N(t) is the number of palindromes and let t represent the number of sequences. \n",
    "The estimator $\\hat{\\lambda}$ was calculated by finding the number of palindromes per sequence and then adding them up and dividing by the number of sequences used.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "695306fe-b451-4da6-b5aa-89207b346166",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Why are palindromes important in DNA sequences?\n",
    "The presence of palindromes indicates that the sequence is a restriction site. A restriction site consists of a short DNA sequence which is recognized by restriction enzymes to cleave the phosphodiester bonds of DNA. I chose to describe a palindromic DNA sequence to be 5 base pairs long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "95996323-7c4e-45be-8c66-5433a03feca1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 3 5-base pair palindromes in the 6th sequences and they occur at indices [8, 16, 34]\n",
      "AGGGGCAAGGAGGATGGAAAGAGGTTGCCGTATAAAGAAACTAGAGTCCGTTTAGGT\n"
     ]
    }
   ],
   "source": [
    "# Define a function to reverse a string\n",
    "def reverse_string(s):\n",
    "    return s[::-1]\n",
    "\n",
    "# Define the window size\n",
    "window_size = 5\n",
    "\n",
    "# Only looks at a single sequence\n",
    "def index_palindrome(seq):\n",
    "    temp = []\n",
    "    # Iterate over the sequence and shift the window one base at a time\n",
    "    for i in range(len(seq) - window_size + 1):\n",
    "        # Look at the 5-base pair sequence\n",
    "        current_sequence = seq[i:i+window_size]\n",
    "        # If the sequence is a palindrome, print the index\n",
    "        if current_sequence == reverse_string(current_sequence):\n",
    "            # If the sequence is a palindrome, print its position in the DNA sequence\n",
    "            temp.append(i) \n",
    "    return temp\n",
    "\n",
    "index_6th_seq = index_palindrome(sequences[5])\n",
    "print('There are', len(index_6th_seq), '5-base pair palindromes in the 6th sequences and they occur at indices', index_6th_seq)\n",
    "print(sequences[5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5b0c38a8-7d3e-45bd-bf5a-7bd7a24b6529",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# palindrome_count will count the number of palindromes present in the sequence\n",
    "palindrome_count = 0\n",
    "arr_palindromes = []\n",
    "\n",
    "for seq in sequences:\n",
    "    # Iterate over the sequence and shift the window one base at a time\n",
    "    for i in range(len(seq) - window_size + 1):\n",
    "        # Look at the 5-base pair sequence\n",
    "        current_sequence = seq[i:i+window_size]\n",
    "        if current_sequence == reverse_string(current_sequence):\n",
    "            # Increase by 1 if the sequence is a palindrome\n",
    "            palindrome_count += 1\n",
    "    arr_palindromes.append(palindrome_count)\n",
    "    palindrome_count = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4107ab6e-9dd4-4955-b654-b8b65d656804",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjIAAAHHCAYAAACle7JuAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/H5lhTAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA6S0lEQVR4nO3deVwV9eL/8fcBZBFZFGVLRUVcUnFNc83CRDQTLSvNxCWzK940l8p7u+JSopVmdd3awMryVi5tV80FtUXzqpFLRkiaGriWIJioML8/+np+nsCF48Fzxl7Px2MeD+ZzZua8Z0B5MzPnHIthGIYAAABMyM3ZAQAAAOxFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQFgKpmZmeratasCAgJksVi0fPny6/K8gwYNUq1atWzGLBaLJk2a5NDnKY9tAjcyigxwCampqbJYLKVOTz31lLPj/WUlJCRo586devbZZ/X222+rVatWpS63f/9+m++Zu7u7atasqd69eys9Pf36hgZQbjycHQBwdVOmTFHt2rVtxho3buykNH9tv//+uzZt2qR//vOfGjly5FWt069fP3Xv3l1FRUXas2eP5s2bpxUrVmjz5s1q1qzZNefx8OC/UcCZ+BcIXEFcXNwl/+r/szNnzsjT01NubpzsLA/Hjh2TJAUGBl71Oi1atNCAAQOs8+3bt9fdd9+tefPmacGCBdeUx9vb+5rWt1dBQYF8fX2d8tyAq+F/W8BO69evl8Vi0eLFi/X000/rpptuUsWKFZWXlydJ+uabb9StWzcFBASoYsWKuu222/TVV1+V2M6XX36pW265Rd7e3oqMjNSCBQs0adIkWSwW6zIXLpOkpqaWWL+0eyp++eUXDRkyRCEhIfLy8lKjRo305ptvlpr//fff17PPPqvq1avL29tbMTEx2rt3b4nn+eabb9S9e3dVrlxZvr6+io6O1ksvvSRJSklJkcVi0bfffltivWnTpsnd3V2//PLLZY/nt99+q7i4OPn7+6tSpUqKiYnR5s2brY9PmjRJERERkqTx48fLYrGUuGflatxxxx2SpH379kmSPvroI/Xo0UPh4eHy8vJSZGSkpk6dqqKioitu68/H/sL3be/evRo0aJACAwMVEBCgwYMH6/Tp0zbrFhYW6vHHH1e1atXk5+enu+++W4cOHSrxHBe2+f3336t///6qXLmyOnToIEk6f/68pk6dqsjISHl5ealWrVr6xz/+ocLCQptt1KpVS3fddZfWr1+vVq1aycfHR02aNNH69eslSUuXLlWTJk3k7e2tli1blvp9/OGHH3TvvfeqSpUq8vb2VqtWrfTxxx/bLHPu3DlNnjxZUVFR8vb2VlBQkDp06KDVq1df8VgC9uKMDHAFubm5On78uM1Y1apVrV9PnTpVnp6eGjdunAoLC+Xp6al169YpLi5OLVu2VFJSktzc3JSSkqI77rhDX3zxhVq3bi1J2rlzp7p27apq1app0qRJOn/+vJKSkhQSEmJ33iNHjujWW2+VxWLRyJEjVa1aNa1YsUJDhw5VXl6eRo8ebbP89OnT5ebmpnHjxik3N1fPPfecHnzwQX3zzTfWZVavXq277rpLYWFhGjVqlEJDQ7Vnzx59+umnGjVqlO69914lJiZq0aJFat68uc32Fy1apM6dO+umm266ZObdu3erY8eO8vf31xNPPKEKFSpowYIF6ty5szZs2KA2bdqoT58+CgwM1OOPP269XFSpUqUyH5+srCxJUlBQkKQ/7oWqVKmSxowZo0qVKmndunWaOHGi8vLy9Pzzz5d5+5J03333qXbt2kpOTtb27dv1+uuvKzg4WDNmzLAu8/DDD+udd95R//791a5dO61bt049evS45Db79u2rqKgoTZs2TYZhWLexcOFC3XvvvRo7dqy++eYbJScna8+ePVq2bJnN+nv37lX//v01fPhwDRgwQC+88IJ69uyp+fPn6x//+IdGjBghSUpOTtZ9992njIwM65nF3bt3q3379rrpppv01FNPydfXV++//77i4+O1ZMkS9e7dW9IfpSs5OVkPP/ywWrdurby8PG3dulXbt2/XnXfeadexBK7IAFCqlJQUQ1Kpk2EYRlpamiHJqFOnjnH69GnresXFxUZUVJQRGxtrFBcXW8dPnz5t1K5d27jzzjutY/Hx8Ya3t7fx888/W8e+//57w93d3bj4n+e+ffsMSUZKSkqJnJKMpKQk6/zQoUONsLAw4/jx4zbLPfDAA0ZAQIA164X8DRs2NAoLC63LvfTSS4YkY+fOnYZhGMb58+eN2rVrGxEREcZvv/1ms82L969fv35GeHi4UVRUZB3bvn37JXNfLD4+3vD09DSysrKsY9nZ2Yafn5/RqVOnEsfh+eefv+z2Ll528uTJxrFjx4zDhw8b69evN5o3b25IMpYsWWIYhmHzvbtg+PDhRsWKFY0zZ85YxxISEoyIiAib5f587JOSkgxJxpAhQ2yW6927txEUFGSdT09PNyQZI0aMsFmuf//+l9xmv379bJa9sI2HH37YZnzcuHGGJGPdunXWsYiICEOS8fXXX1vHVq1aZUgyfHx8bH7+FixYYEgy0tLSrGMxMTFGkyZNbI5HcXGx0a5dOyMqKso61rRpU6NHjx4GcD1xaQm4gjlz5mj16tU208USEhLk4+NjnU9PT1dmZqb69++vEydO6Pjx4zp+/LgKCgoUExOjjRs3qri4WEVFRVq1apXi4+NVs2ZN6/oNGzZUbGysXVkNw9CSJUvUs2dPGYZhfe7jx48rNjZWubm52r59u806gwcPlqenp3W+Y8eOkqSffvpJ0h+XfPbt26fRo0eXuDfl4stfAwcOVHZ2ttLS0qxjixYtko+Pj+65555LZi4qKtLnn3+u+Ph41alTxzoeFham/v3768svv7RerrNHUlKSqlWrptDQUHXu3FlZWVmaMWOG+vTpI0k237tTp07p+PHj6tixo06fPq0ffvjBrud89NFHbeY7duyoEydOWPfjv//9ryTpscces1nuz2fLLrfNC9sYM2aMzfjYsWMlSZ999pnN+M0336y2bdta59u0aSPpj0ttF//8XRi/8P3/9ddftW7dOt13333W43P8+HGdOHFCsbGxyszMtF42DAwM1O7du5WZmXnJ/QAcjUtLwBW0bt36sjf7/vkVTRf+E09ISLjkOrm5uSosLNTvv/+uqKioEo/Xr1/f+ouqLI4dO6aTJ0/q1Vdf1auvvlrqMkePHrWZv/iXmCRVrlxZkvTbb79J+v+XYq70Sq0777xTYWFhWrRokWJiYlRcXKz33ntPvXr1kp+f32Uznz59WvXr1y/xWMOGDVVcXKyDBw+qUaNGl33+S3nkkUfUt29fubm5KTAwUI0aNZKXl5f18d27d+vpp5/WunXrShSm3Nxcu57zcsfU399fP//8s9zc3BQZGWmzXGnH4II//5xd2EbdunVtxkNDQxUYGKiff/75spkCAgIkSTVq1Ch1/ML3f+/evTIMQ//617/0r3/9q9RsR48e1U033aQpU6aoV69eqlevnho3bqxu3brpoYceUnR09CX3C7hWFBngGl38F70kFRcXS5Kef/75S768t1KlSiVuyLyci898XOzPN6ReeO4BAwZcskj9+ZeKu7t7qcsZ/3cfxtVyd3dX//799dprr2nu3Ln66quvlJ2dbfOKIWeIiopSly5dSn3s5MmTuu222+Tv768pU6YoMjJS3t7e2r59u5588knr8SwrRx3Ti/355+yCS/1sXG2mK2W9cAzGjRt3yTOFF8pUp06dlJWVpY8++kiff/65Xn/9db344ouaP3++Hn744avKCZQVRQZwsAt/Zfv7+1/yF6gkVatWTT4+PqWehs/IyLCZv/AX/cmTJ23G//xX94VXwBQVFV32ucviwv7s2rXritscOHCgZs6cqU8++UQrVqxQtWrVrniZrFq1aqpYsWKJfZb+eKWMm5tbibMGjrJ+/XqdOHFCS5cuVadOnazjF17RVF4iIiJUXFysrKwsm7MwpR2DK20jMzNTDRs2tI4fOXJEJ0+etL7C61pduNxXoUKFq/qZqlKligYPHqzBgwcrPz9fnTp10qRJkygyKDfcIwM4WMuWLRUZGakXXnhB+fn5JR6/8F4o7u7uio2N1fLly3XgwAHr43v27NGqVats1vH391fVqlW1ceNGm/G5c+fazLu7u+uee+7RkiVLtGvXrks+d1m0aNFCtWvX1uzZs0sUqT+fYYiOjlZ0dLRef/11LVmyRA888MAV3zDO3d1dXbt21UcffaT9+/dbx48cOaJ3331XHTp0kL+/f5lzX40LZyMu3o+zZ8+WOK6OFhcXJ0l6+eWXbcZnz5591dvo3r17qevMmjVLki77CqiyCA4OVufOnbVgwQLl5OSUePzin6kTJ07YPFapUiXVrVu3TGcfgbLijAzgYG5ubnr99dcVFxenRo0aafDgwbrpppv0yy+/KC0tTf7+/vrkk08kSZMnT9bKlSvVsWNHjRgxQufPn9crr7yiRo0aaceOHTbbffjhhzV9+nQ9/PDDatWqlTZu3Kgff/yxxPNPnz5daWlpatOmjYYNG6abb75Zv/76q7Zv3641a9bo119/LfP+zJs3Tz179lSzZs00ePBghYWF6YcfftDu3btLlK6BAwdq3LhxknTVl5WeeeYZrV69Wh06dNCIESPk4eGhBQsWqLCwUM8991yZ8pZFu3btVLlyZSUkJOixxx6TxWLR22+/fU2XgK5Gs2bN1K9fP82dO1e5ublq166d1q5dW+r791xK06ZNlZCQoFdffdV6iWzLli1auHCh4uPjdfvttzss75w5c9ShQwc1adJEw4YNU506dXTkyBFt2rRJhw4d0nfffSfpjxuKO3furJYtW6pKlSraunWrPvzww6t+F2bAHhQZoBx07txZmzZt0tSpU/Xvf/9b+fn5Cg0NVZs2bTR8+HDrctHR0Vq1apXGjBmjiRMnqnr16po8ebJycnJKFJmJEyfq2LFj+vDDD/X+++8rLi5OK1asUHBwsM1yISEh2rJli6ZMmaKlS5dq7ty5CgoKUqNGjWzex6QsYmNjlZaWpsmTJ2vmzJkqLi5WZGSkhg0bVmLZBx98UE8++aQiIyOt75dzJY0aNdIXX3yhCRMmKDk5WcXFxWrTpo3eeecd66toykNQUJA+/fRTjR07Vk8//bQqV66sAQMGKCYmxu5Xjl2tN998U9WqVdOiRYu0fPly3XHHHfrss8/KdBnt9ddfV506dZSamqply5YpNDRUEyZMUFJSkkOz3nzzzdq6dasmT56s1NRUnThxQsHBwWrevLkmTpxoXe6xxx7Txx9/rM8//1yFhYWKiIjQM888o/Hjxzs0D3Axi1Hef3oAKLNJkyZp8uTJ5X5moDwcP35cYWFhmjhx4iVf5QIAjsI9MgAcKjU1VUVFRXrooYecHQXAXwCXlgA4xLp16/T999/r2WefVXx8vF2fgwQAZUWRAeAQU6ZM0ddff6327dvrlVdecXYcAH8R3CMDAABMi3tkAACAaVFkAACAad3w98gUFxcrOztbfn5+V/2ZJAAAwLkMw9CpU6cUHh4uN7dLn3e54YtMdnZ2uX1OCwAAKF8HDx5U9erVL/n4DV9k/Pz8JP1xIMrr81oAlIOCAik8/I+vs7MlX1/n5gFwXeXl5alGjRrW3+OXcsMXmQuXk/z9/SkygJn83wc6SpL8/SkywF/UlW4L4WZfAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWk4tMsnJybrlllvk5+en4OBgxcfHKyMjw2aZzp07y2Kx2EyPPvqokxIDAABX4tQis2HDBiUmJmrz5s1avXq1zp07p65du6qgoMBmuWHDhiknJ8c6Pffcc05KDAAAXIlTPzRy5cqVNvOpqakKDg7Wtm3b1KlTJ+t4xYoVFRoaer3jAQAAF+dS98jk5uZKkqpUqWIzvmjRIlWtWlWNGzfWhAkTdPr0aWfEAwAALsapZ2QuVlxcrNGjR6t9+/Zq3Lixdbx///6KiIhQeHi4duzYoSeffFIZGRlaunRpqdspLCxUYWGhdT4vL6/cswMAAOdwmSKTmJioXbt26csvv7QZf+SRR6xfN2nSRGFhYYqJiVFWVpYiIyNLbCc5OVmTJ08u97wAXFutpz5zdgS77J/ew9kRAFNxiUtLI0eO1Keffqq0tDRVr179ssu2adNGkrR3795SH58wYYJyc3Ot08GDBx2eFwAAuAannpExDEN///vftWzZMq1fv161a9e+4jrp6emSpLCwsFIf9/LykpeXlyNjAgAAF+XUIpOYmKh3331XH330kfz8/HT48GFJUkBAgHx8fJSVlaV3331X3bt3V1BQkHbs2KHHH39cnTp1UnR0tDOjAwAAF+DUIjNv3jxJf7zp3cVSUlI0aNAgeXp6as2aNZo9e7YKCgpUo0YN3XPPPXr66aedkBYAALgap19aupwaNWpow4YN1ykNAAAwG5e42RcAAMAeFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaFBkAAGBaHs4OALiKWk995uwIdtk/vYezIwCA03BGBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmJZTi0xycrJuueUW+fn5KTg4WPHx8crIyLBZ5syZM0pMTFRQUJAqVaqke+65R0eOHHFSYgAA4EqcWmQ2bNigxMREbd68WatXr9a5c+fUtWtXFRQUWJd5/PHH9cknn+iDDz7Qhg0blJ2drT59+jgxNQAAcBUeznzylStX2synpqYqODhY27ZtU6dOnZSbm6s33nhD7777ru644w5JUkpKiho2bKjNmzfr1ltvdUZsAADgIlzqHpnc3FxJUpUqVSRJ27Zt07lz59SlSxfrMg0aNFDNmjW1adOmUrdRWFiovLw8mwkAANyYXKbIFBcXa/To0Wrfvr0aN24sSTp8+LA8PT0VGBhos2xISIgOHz5c6naSk5MVEBBgnWrUqFHe0QEAgJO4TJFJTEzUrl27tHjx4mvazoQJE5Sbm2udDh486KCEAADA1Tj1HpkLRo4cqU8//VQbN25U9erVreOhoaE6e/asTp48aXNW5siRIwoNDS11W15eXvLy8irvyAAAwAU49YyMYRgaOXKkli1bpnXr1ql27do2j7ds2VIVKlTQ2rVrrWMZGRk6cOCA2rZte73jAgAAF+PUMzKJiYl699139dFHH8nPz89630tAQIB8fHwUEBCgoUOHasyYMapSpYr8/f3197//XW3btuUVSwAAwLlFZt68eZKkzp0724ynpKRo0KBBkqQXX3xRbm5uuueee1RYWKjY2FjNnTv3OicFAACuyKlFxjCMKy7j7e2tOXPmaM6cOdchEQAAMBOXedUSAABAWVFkAACAaVFkAACAaVFkAACAaVFkAACAaVFkAACAaVFkAACAabnEZy0BuH5qPfWZsyNcFZ+zZ7Tn/75u+K+V2jPrHqfmAeCaOCMDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMiyIDAABMy8PZAQAA9qn11GfOjmC3/dN7ODsCbhCckQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZFkQEAAKZlV5H56aefHJ0DAACgzOwqMnXr1tXtt9+ud955R2fOnHF0JgAAgKtiV5HZvn27oqOjNWbMGIWGhmr48OHasmVLmbezceNG9ezZU+Hh4bJYLFq+fLnN44MGDZLFYrGZunXrZk9kAABwA7KryDRr1kwvvfSSsrOz9eabbyonJ0cdOnRQ48aNNWvWLB07duyqtlNQUKCmTZtqzpw5l1ymW7duysnJsU7vvfeePZEBAMAN6Jpu9vXw8FCfPn30wQcfaMaMGdq7d6/GjRunGjVqaODAgcrJybns+nFxcXrmmWfUu3fvSy7j5eWl0NBQ61S5cuVriQwAAG4g11Rktm7dqhEjRigsLEyzZs3SuHHjlJWVpdWrVys7O1u9evW65oDr169XcHCw6tevr7/97W86ceLENW8TAADcGOz69OtZs2YpJSVFGRkZ6t69u9566y11795dbm5/9KLatWsrNTVVtWrVuqZw3bp1U58+fVS7dm1lZWXpH//4h+Li4rRp0ya5u7uXuk5hYaEKCwut83l5edeUAQAAuC67isy8efM0ZMgQDRo0SGFhYaUuExwcrDfeeOOawj3wwAPWr5s0aaLo6GhFRkZq/fr1iomJKXWd5ORkTZ48+ZqeFwAAmINdl5YyMzM1YcKES5YYSfL09FRCQoLdwUpTp04dVa1aVXv37r3kMhMmTFBubq51OnjwoEMzAAAA12HXGZmUlBRVqlRJffv2tRn/4IMPdPr0aYcXmAsOHTqkEydOXLZAeXl5ycvLq1yeHwAAuBa7zsgkJyeratWqJcaDg4M1bdq0q95Ofn6+0tPTlZ6eLknat2+f0tPTdeDAAeXn52v8+PHavHmz9u/fr7Vr16pXr16qW7euYmNj7YkNAABuMHadkTlw4IBq165dYjwiIkIHDhy46u1s3bpVt99+u3V+zJgxkqSEhATNmzdPO3bs0MKFC3Xy5EmFh4era9eumjp1KmdcAACAJDuLTHBwsHbs2FHiVUnfffedgoKCrno7nTt3lmEYl3x81apV9sQDAAB/EXZdWurXr58ee+wxpaWlqaioSEVFRVq3bp1GjRpl80ojAACA8mTXGZmpU6dq//79iomJkYfHH5soLi7WwIEDy3SPDAAAwLWwq8h4enrqP//5j6ZOnarvvvtOPj4+atKkiSIiIhydDwAA4JLsKjIX1KtXT/Xq1XNUFgAAgDKxq8gUFRUpNTVVa9eu1dGjR1VcXGzz+Lp16xwSDgAA4HLsKjKjRo1SamqqevToocaNG8tisTg6FwAAwBXZVWQWL16s999/X927d3d0HgAAgKtm18uvPT09VbduXUdnAQAAKBO7iszYsWP10ksvXfbN7AAAAMqbXZeWvvzyS6WlpWnFihVq1KiRKlSoYPP40qVLHRIOAADgcuwqMoGBgerdu7ejswAAAJSJXUUmJSXF0TkAAADKzK57ZCTp/PnzWrNmjRYsWKBTp05JkrKzs5Wfn++wcAAAAJdj1xmZn3/+Wd26ddOBAwdUWFioO++8U35+fpoxY4YKCws1f/58R+cEAAAowa4zMqNGjVKrVq3022+/ycfHxzreu3dvrV271mHhAAAALseuMzJffPGFvv76a3l6etqM16pVS7/88otDggEAAFyJXWdkiouLVVRUVGL80KFD8vPzu+ZQAAAAV8OuItO1a1fNnj3bOm+xWJSfn6+kpCQ+tgAAAFw3dl1amjlzpmJjY3XzzTfrzJkz6t+/vzIzM1W1alW99957js4IAABQKruKTPXq1fXdd99p8eLF2rFjh/Lz8zV06FA9+OCDNjf/AgAAlCe7iowkeXh4aMCAAY7MAgAAUCZ2FZm33nrrso8PHDjQrjAAAABlYVeRGTVqlM38uXPndPr0aXl6eqpixYoUGQAAcF3Y9aql3377zWbKz89XRkaGOnTowM2+AADgurH7s5b+LCoqStOnTy9xtgYAAKC8OKzISH/cAJydne3ITQIAAFySXffIfPzxxzbzhmEoJydH//73v9W+fXuHBAMAALgSu4pMfHy8zbzFYlG1atV0xx13aObMmY7IBQAAcEV2FZni4mJH5wAAACgzh94jAwAAcD3ZdUZmzJgxV73srFmz7HkKAACAK7KryHz77bf69ttvde7cOdWvX1+S9OOPP8rd3V0tWrSwLmexWByTEgAAoBR2FZmePXvKz89PCxcuVOXKlSX98SZ5gwcPVseOHTV27FiHhgQAACiNXffIzJw5U8nJydYSI0mVK1fWM888w6uWAADAdWNXkcnLy9OxY8dKjB87dkynTp265lAAAABXw64i07t3bw0ePFhLly7VoUOHdOjQIS1ZskRDhw5Vnz59HJ0RAACgVHbdIzN//nyNGzdO/fv317lz5/7YkIeHhg4dqueff96hAQEAAC7FriJTsWJFzZ07V88//7yysrIkSZGRkfL19XVoOAAAgMu5pjfEy8nJUU5OjqKiouTr6yvDMByVCwAA4IrsKjInTpxQTEyM6tWrp+7duysnJ0eSNHToUF56DQAArhu7iszjjz+uChUq6MCBA6pYsaJ1/P7779fKlSsdFg4AAOBy7LpH5vPPP9eqVatUvXp1m/GoqCj9/PPPDgkGAABwJXadkSkoKLA5E3PBr7/+Ki8vr2sOBQAAcDXsKjIdO3bUW2+9ZZ23WCwqLi7Wc889p9tvv91h4QAAAC7HrktLzz33nGJiYrR161adPXtWTzzxhHbv3q1ff/1VX331laMzAgAAlMquMzKNGzfWjz/+qA4dOqhXr14qKChQnz599O233yoyMtLRGQEAAEpV5jMy586dU7du3TR//nz985//LI9MAAAAV6XMZ2QqVKigHTt2lEcWAACAMrHr0tKAAQP0xhtvODoLAABAmdh1s+/58+f15ptvas2aNWrZsmWJz1iaNWuWQ8IBAABcTpmKzE8//aRatWpp165datGihSTpxx9/tFnGYrE4Lh1cQq2nPnN2BLvtn97D2REAAOWoTEUmKipKOTk5SktLk/THRxK8/PLLCgkJKZdwAAAAl1Ome2T+/OnWK1asUEFBgUMDAQAAXC27bva94M/FBgAA4HoqU5GxWCwl7oHhnhgAAOAsZbpHxjAMDRo0yPrBkGfOnNGjjz5a4lVLS5cudVxCAACASyhTkUlISLCZHzBggEPDAAAAlEWZikxKSkp55QAAACiza7rZFwAAwJkoMgAAwLScWmQ2btyonj17Kjw8XBaLRcuXL7d53DAMTZw4UWFhYfLx8VGXLl2UmZnpnLAAAMDlOLXIFBQUqGnTppozZ06pjz/33HN6+eWXNX/+fH3zzTfy9fVVbGyszpw5c52TAgAAV2TXh0Y6SlxcnOLi4kp9zDAMzZ49W08//bR69eolSXrrrbcUEhKi5cuX64EHHrieUQEAgAty2Xtk9u3bp8OHD6tLly7WsYCAALVp00abNm265HqFhYXKy8uzmQAAwI3JZYvM4cOHJanEB1KGhIRYHytNcnKyAgICrFONGjXKNScAAHAely0y9powYYJyc3Ot08GDB50dCQAAlBOXLTKhoaGSpCNHjtiMHzlyxPpYaby8vOTv728zAQCAG5PLFpnatWsrNDRUa9eutY7l5eXpm2++Udu2bZ2YDAAAuAqnvmopPz9fe/futc7v27dP6enpqlKlimrWrKnRo0frmWeeUVRUlGrXrq1//etfCg8PV3x8vPNCAwAAl+HUIrN161bdfvvt1vkxY8ZI+uPDKVNTU/XEE0+ooKBAjzzyiE6ePKkOHTpo5cqV8vb2dlZkAADgQpxaZDp37izDMC75uMVi0ZQpUzRlypTrmAoAAJiFy94jAwAAcCUUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoUGQAAYFoezg4AAMDl1HrqM2dHsNv+6T2cHeGGxxkZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWhQZAABgWi5dZCZNmiSLxWIzNWjQwNmxAACAi/BwdoAradSokdasWWOd9/Bw+cgAAOA6cflW4OHhodDQUGfHAAAALsilLy1JUmZmpsLDw1WnTh09+OCDOnDgwGWXLywsVF5ens0EAABuTC5dZNq0aaPU1FStXLlS8+bN0759+9SxY0edOnXqkuskJycrICDAOtWoUeM6JgYAANeTSxeZuLg49e3bV9HR0YqNjdV///tfnTx5Uu+///4l15kwYYJyc3Ot08GDB69jYgAAcD25/D0yFwsMDFS9evW0d+/eSy7j5eUlLy+v65gKAAA4i0ufkfmz/Px8ZWVlKSwszNlRAACAC3DpIjNu3Dht2LBB+/fv19dff63evXvL3d1d/fr1c3Y0AADgAlz60tKhQ4fUr18/nThxQtWqVVOHDh20efNmVatWzdnRAACAC3DpIrN48WJnRwAAAC7MpS8tAQAAXA5FBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmBZFBgAAmJaHswMAAACp1lOfOTuCXfZP7+HU5+eMDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC0PZwcwM7N+5Lrk/I9dBwDAETgjAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATIsiAwAATMsURWbOnDmqVauWvL291aZNG23ZssXZkQAAgAtw+SLzn//8R2PGjFFSUpK2b9+upk2bKjY2VkePHnV2NAAA4GQuX2RmzZqlYcOGafDgwbr55ps1f/58VaxYUW+++aazowEAACdz6SJz9uxZbdu2TV26dLGOubm5qUuXLtq0aZMTkwEAAFfg4ewAl3P8+HEVFRUpJCTEZjwkJEQ//PBDqesUFhaqsLDQOp+bmytJysvLc3i+4sLTDt/m9VKW48F+uray/mybZT+Lzp7RhT0rKjx9w+7nn/0VfmYl9rM0Zt3P8vj9evF2DcO4/IKGC/vll18MScbXX39tMz5+/HijdevWpa6TlJRkSGJiYmJiYmK6AaaDBw9etiu49BmZqlWryt3dXUeOHLEZP3LkiEJDQ0tdZ8KECRozZox1vri4WL/++quCgoJksVjKNa8j5eXlqUaNGjp48KD8/f2dHadc/BX2UWI/bzTs543jr7CPknn30zAMnTp1SuHh4ZddzqWLjKenp1q2bKm1a9cqPj5e0h/FZO3atRo5cmSp63h5ecnLy8tmLDAwsJyTlh9/f39T/eDZ46+wjxL7eaNhP28cf4V9lMy5nwEBAVdcxqWLjCSNGTNGCQkJatWqlVq3bq3Zs2eroKBAgwcPdnY0AADgZC5fZO6//34dO3ZMEydO1OHDh9WsWTOtXLmyxA3AAADgr8fli4wkjRw58pKXkm5UXl5eSkpKKnGZ7EbyV9hHif280bCfN46/wj5KN/5+WgzjSq9rAgAAcE0u/YZ4AAAAl0ORAQAApkWRAQAApkWRAQAApkWRcUFz5sxRrVq15O3trTZt2mjLli3OjuRwGzduVM+ePRUeHi6LxaLly5c7O5LDJScn65ZbbpGfn5+Cg4MVHx+vjIwMZ8dyuHnz5ik6Otr6Zltt27bVihUrnB2rXE2fPl0Wi0WjR492dhSHmjRpkiwWi83UoEEDZ8cqF7/88osGDBigoKAg+fj4qEmTJtq6dauzYzlUrVq1Snw/LRaLEhMTnR3NoSgyLuY///mPxowZo6SkJG3fvl1NmzZVbGysjh496uxoDlVQUKCmTZtqzpw5zo5SbjZs2KDExERt3rxZq1ev1rlz59S1a1cVFBQ4O5pDVa9eXdOnT9e2bdu0detW3XHHHerVq5d2797t7Gjl4n//+58WLFig6OhoZ0cpF40aNVJOTo51+vLLL50dyeF+++03tW/fXhUqVNCKFSv0/fffa+bMmapcubKzoznU//73P5vv5erVqyVJffv2dXIyB3PMxzvCUVq3bm0kJiZa54uKiozw8HAjOTnZianKlyRj2bJlzo5R7o4ePWpIMjZs2ODsKOWucuXKxuuvv+7sGA536tQpIyoqyli9erVx2223GaNGjXJ2JIdKSkoymjZt6uwY5e7JJ580OnTo4OwY192oUaOMyMhIo7i42NlRHIozMi7k7Nmz2rZtm7p06WIdc3NzU5cuXbRp0yYnJoMj5ObmSpKqVKni5CTlp6ioSIsXL1ZBQYHatm3r7DgOl5iYqB49etj8G73RZGZmKjw8XHXq1NGDDz6oAwcOODuSw3388cdq1aqV+vbtq+DgYDVv3lyvvfaas2OVq7Nnz+qdd97RkCFDTPUByleDIuNCjh8/rqKiohIfvxASEqLDhw87KRUcobi4WKNHj1b79u3VuHFjZ8dxuJ07d6pSpUry8vLSo48+qmXLlunmm292diyHWrx4sbZv367k5GRnRyk3bdq0UWpqqlauXKl58+Zp37596tixo06dOuXsaA71008/ad68eYqKitKqVav0t7/9TY899pgWLlzo7GjlZvny5Tp58qQGDRrk7CgOZ4qPKADMLjExUbt27boh7zeQpPr16ys9PV25ubn68MMPlZCQoA0bNtwwZebgwYMaNWqUVq9eLW9vb2fHKTdxcXHWr6Ojo9WmTRtFRETo/fff19ChQ52YzLGKi4vVqlUrTZs2TZLUvHlz7dq1S/Pnz1dCQoKT05WPN954Q3FxcQoPD3d2FIfjjIwLqVq1qtzd3XXkyBGb8SNHjig0NNRJqXCtRo4cqU8//VRpaWmqXr26s+OUC09PT9WtW1ctW7ZUcnKymjZtqpdeesnZsRxm27ZtOnr0qFq0aCEPDw95eHhow4YNevnll+Xh4aGioiJnRywXgYGBqlevnvbu3evsKA4VFhZWomQ3bNjwhryMJkk///yz1qxZo4cfftjZUcoFRcaFeHp6qmXLllq7dq11rLi4WGvXrr0h7ze40RmGoZEjR2rZsmVat26dateu7exI101xcbEKCwudHcNhYmJitHPnTqWnp1unVq1a6cEHH1R6errc3d2dHbFc5OfnKysrS2FhYc6O4lDt27cv8VYIP/74oyIiIpyUqHylpKQoODhYPXr0cHaUcsGlJRczZswYJSQkqFWrVmrdurVmz56tgoICDR482NnRHCo/P9/mr7x9+/YpPT1dVapUUc2aNZ2YzHESExP17rvv6qOPPpKfn5/1PqeAgAD5+Pg4OZ3jTJgwQXFxcapZs6ZOnTqld999V+vXr9eqVaucHc1h/Pz8Stzb5Ovrq6CgoBvqnqdx48apZ8+eioiIUHZ2tpKSkuTu7q5+/fo5O5pDPf7442rXrp2mTZum++67T1u2bNGrr76qV1991dnRHK64uFgpKSlKSEiQh8cN+ivf2S+bQkmvvPKKUbNmTcPT09No3bq1sXnzZmdHcri0tDRDUokpISHB2dEcprT9k2SkpKQ4O5pDDRkyxIiIiDA8PT2NatWqGTExMcbnn3/u7Fjl7kZ8+fX9999vhIWFGZ6ensZNN91k3H///cbevXudHatcfPLJJ0bjxo0NLy8vo0GDBsarr77q7EjlYtWqVYYkIyMjw9lRyo3FMAzDORUKAADg2nCPDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDAAAMC2KDPAXtn//flksFqWnpzs7itUPP/ygW2+9Vd7e3mrWrFm5PEdqaqoCAwOt85MmTXLIc9WqVUuzZ8++5u0AuHoUGcCJBg0aJIvFounTp9uML1++XBaLxUmpnCspKUm+vr7KyMiw+dyxi104bhaLxfqBlVOmTNH58+ftes5x48Zd8rkAuDaKDOBk3t7emjFjhn777TdnR3GYs2fP2r1uVlaWOnTooIiICAUFBV1yuW7duiknJ0eZmZkaO3asJk2apOeff96u56xUqdJln8tRzp07V+7PAfzVUGQAJ+vSpYtCQ0OVnJx8yWVKu/Qxe/Zs1apVyzo/aNAgxcfHa9q0aQoJCVFgYKD1LMX48eNVpUoVVa9eXSkpKSW2/8MPP6hdu3by9vZW48aNtWHDBpvHd+3apbi4OFWqVEkhISF66KGHdPz4cevjnTt31siRIzV69GhVrVpVsbGxpe5HcXGxpkyZourVq8vLy0vNmjXTypUrrY9bLBZt27ZNU6ZMkcVi0aRJky55TLy8vBQaGqqIiAj97W9/U5cuXfTxxx9LkmbNmqUmTZrI19dXNWrU0IgRI5Sfn3/Jbf35+F44li+88ILCwsIUFBSkxMREmyJy9OhR9ezZUz4+Pqpdu7YWLVpUYrsWi0Xz5s3T3XffLV9fXz377LOSpHnz5ikyMlKenp6qX7++3n777RLrLViwQHfddZcqVqyohg0batOmTdq7d686d+4sX19ftWvXTllZWTbrffTRR2rRooW8vb1Vp04dTZ482XqWyjAMTZo0STVr1pSXl5fCw8P12GOPXfKYAGZBkQGczN3dXdOmTdMrr7yiQ4cOXdO21q1bp+zsbG3cuFGzZs1SUlKS7rrrLlWuXFnffPONHn30UQ0fPrzE84wfP15jx47Vt99+q7Zt26pnz546ceKEJOnkyZO644471Lx5c23dulUrV67UkSNHdN9999lsY+HChfL09NRXX32l+fPnl5rvpZde0syZM/XCCy9ox44dio2N1d13363MzExJUk5Ojho1aqSxY8cqJydH48aNu+p99/HxsZ4JcnNz08svv6zdu3dr4cKFWrdunZ544omr3pYkpaWlKSsrS2lpaVq4cKFSU1OVmppqfXzQoEE6ePCg0tLS9OGHH2ru3Lk6evRoie1MmjRJvXv31s6dOzVkyBAtW7ZMo0aN0tixY7Vr1y4NHz5cgwcPVlpams16U6dO1cCBA5Wenq4GDRqof//+Gj58uCZMmKCtW7fKMAyNHDnSuvwXX3yhgQMHatSoUfr++++1YMECpaamWsvTkiVL9OKLL2rBggXKzMzU8uXL1aRJkzIdE8AlOfczK4G/toSEBKNXr16GYRjGrbfeagwZMsQwDMNYtmyZcfE/z6SkJKNp06Y267744otGRESEzbYiIiKMoqIi61j9+vWNjh07WufPnz9v+Pr6Gu+9955hGIaxb98+Q5Ixffp06zLnzp0zqlevbsyYMcMwDMOYOnWq0bVrV5vnPnjwoM0n6t52221G8+bNr7i/4eHhxrPPPmszdssttxgjRoywzjdt2tRISkq67HYuPm7FxcXG6tWrDS8vL2PcuHGlLv/BBx8YQUFB1vmUlBQjICDAOv/n43vhWJ4/f9461rdvX+P+++83DMMwMjIyDEnGli1brI/v2bPHkGS8+OKL1jFJxujRo22ytGvXzhg2bJjNWN++fY3u3bvbrPf0009b5zdt2mRIMt544w3r2HvvvWd4e3tb52NiYoxp06bZbPftt982wsLCDMMwjJkzZxr16tUzzp49W/IAASbGGRnARcyYMUMLFy7Unj177N5Go0aN5Ob2//9Zh4SE2PzV7e7urqCgoBJnDtq2bWv92sPDQ61atbLm+O6775SWlqZKlSpZpwYNGkiSzaWNli1bXjZbXl6esrOz1b59e5vx9u3b27XPn376qSpVqiRvb2/FxcXp/vvvt16KWrNmjWJiYnTTTTfJz89PDz30kE6cOKHTp09f9fYbNWokd3d363xYWJj1uO3Zs0ceHh42+9ygQQObV0Jd0KpVK5v5PXv2XNUxiI6Otn4dEhIiSTbfy5CQEJ05c0Z5eXmS/vg+TZkyxeb7NGzYMOXk5Oj06dPq27evfv/9d9WpU0fDhg3TsmXL7L45GnAlFBnARXTq1EmxsbGaMGFCicfc3NxkGIbNWGk3jlaoUMFm3mKxlDpWXFx81bny8/PVs2dPpaen20yZmZnq1KmTdTlfX9+r3qYj3H777dYcv//+uxYuXChfX1/t379fd911l6Kjo7VkyRJt27ZNc+bMkVS2m5Cv9bhdYO9xufj5L7yCrbSxC5ny8/M1efJkm+/Rzp07lZmZKW9vb9WoUUMZGRmaO3eufHx8NGLECHXq1IkbkGF6FBnAhUyfPl2ffPKJNm3aZDNerVo1HT582KbMOPK9XzZv3mz9+vz589q2bZsaNmwoSWrRooV2796tWrVqqW7dujZTWX5J+/v7Kzw8XF999ZXN+FdffaWbb765zJl9fX1Vt25d1axZUx4eHtbxbdu2qbi4WDNnztStt96qevXqKTs7u8zbv5wGDRpYj9MFGRkZOnny5BXXbdiwocOOwcVatGihjIyMEt+junXrWs/S+fj4qGfPnnr55Ze1fv16bdq0STt37rym5wWczePKiwC4Xpo0aaIHH3xQL7/8ss14586ddezYMT333HO69957tXLlSq1YsUL+/v4Oed45c+YoKipKDRs21IsvvqjffvtNQ4YMkSQlJibqtddeU79+/fTEE0+oSpUq2rt3rxYvXqzXX3/d5vLLlYwfP15JSUmKjIxUs2bNlJKSovT09FJf8WOvunXr6ty5c3rllVfUs2fPy958bK/69eurW7duGj58uObNmycPDw+NHj1aPj4+V1x3/Pjxuu+++9S8eXN16dJFn3zyiZYuXao1a9ZcU6aJEyfqrrvuUs2aNXXvvffKzc1N3333nXbt2qVnnnlGqampKioqUps2bVSxYkW988478vHxUURExDU9L+BsnJEBXMyUKVNKXMJo2LCh5s6dqzlz5qhp06basmVLmV7RcyXTp0/X9OnT1bRpU3355Zf6+OOPVbVqVUmynkUpKipS165d1aRJE40ePVqBgYE29+Ncjccee0xjxozR2LFj1aRJE61cuVIff/yxoqKiHLYvTZs21axZszRjxgw1btxYixYtuuxL2+2VkpKi8PBw3XbbberTp48eeeQRBQcHX3G9+Ph4vfTSS3rhhRfUqFEjLViwQCkpKercufM15YmNjdWnn36qzz//XLfccotuvfVWvfjii9aiEhgYqNdee03t27dXdHS01qxZo08++eS6vH8OUJ4sxp8vvAMAAJgEZ2QAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBpUWQAAIBp/T8RuoXZmsE3FgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mean_pal = np.mean(arr_palindromes)\n",
    "\n",
    "# Get the unique values and their counts\n",
    "values, counts = np.unique(arr_palindromes, return_counts=True)\n",
    "\n",
    "# Plot the frequency of numbers\n",
    "plt.bar(values, counts)\n",
    "plt.xlabel('Number of Palindromes')\n",
    "plt.ylabel('Frequency')\n",
    "plt.title('Frequency of Palindromes')\n",
    "plt.axvline(x=mean_pal, color='red')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8418b91b-b990-4b82-95f3-c9b0cfff54e7",
   "metadata": {},
   "source": [
    "The figure above has a very close Gaussian distribution when observing the frequency of palindromes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e33f97fd-9ce2-4cba-a516-6d0fb11c8522",
   "metadata": {},
   "source": [
    "#### What is the probability that there are 10 palindromes detected within the first 10 DNA sequences? <br>What about the probability of 3 palindromes within the first 10 DNA sequences?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "45688e78-5690-450a-8c3c-26d857115b3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P(N(10) = 10) = 1.222221993695921e-06\n",
      "P(N(10) = 3) = 1.5039223762689678e-11\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "P_10_palindromes = ( np.exp(-mean_pal*10) * (mean_pal*10)**10 ) / math.factorial(10)\n",
    "print('P(N(10) = 10) =', P_10_palindromes)\n",
    "P_3_palindromes = ( np.exp(-mean_pal*10) * (mean_pal*10)**3 ) / math.factorial(3)\n",
    "print('P(N(10) = 3) =', P_3_palindromes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0103c500-ec2f-4d31-8d52-8afa0a38c551",
   "metadata": {},
   "source": [
    "Using the assumption of Compound Poisson Distribution:<br>\n",
    "$\\lambda = w \\theta^L$\n",
    "<br>w is the window length.\n",
    "<br>$\\theta = 2 \\cdot (p_A \\cdot p_T + p_C \\cdot p_G)$\n",
    "<br>L is minimal palindrome length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "98d477a7-881b-4669-b4ad-83b970949499",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using the Compound Poisson Distribution, lambda is calculated to be 0.246\n"
     ]
    }
   ],
   "source": [
    "steady_state_seqs = [0.261, 0.228, 0.229, 0.282]\n",
    "theta = 2* ( steady_state_seqs[0]*steady_state_seqs[1] + steady_state_seqs[2]*steady_state_seqs[3] )\n",
    "lamb = 4*theta**2\n",
    "print('Using the Compound Poisson Distribution, lambda is calculated to be', round(lamb, 3))\n",
    "\n",
    "# http://www.math.utep.edu/faculty/mleung/presentation/Leung-JMM100115Slides.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a029c47b-ae4c-4264-a8db-bd37a16bbdee",
   "metadata": {},
   "source": [
    "The Compound Poisson Distribution was used to compare the calculated lambda from the Compound Poisson Distribution to the calculated lambda from the histogram. It is clearly shown that there is a very large difference between the two values which shows how important it is to make correct assumptions for your model. I used the simple mean of the counts of palindromes of each sequence which worked well for my model."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38fffa35-3adc-4f61-a5ae-73a8e215f7de",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Brownian Motion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7a03a41-7884-4a58-b1e6-f3f3e72da1c0",
   "metadata": {},
   "source": [
    "### This Brownian motion simulator below simulates the time it takes to detect the first occurrence of a value using Arbitrage Theorem.\n",
    "The simulation can be applied to many applications where you want to find how long it takes for an even to occur."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf5309fb-2bef-447e-bb2e-c18abc8e5607",
   "metadata": {},
   "source": [
    "Geometric Brownian Motion\n",
    "$S_t=S_0\\exp\\bigg((r-\\frac{1}{2}\\sigma^2)t+\\sigma\\sqrt{t}z\\bigg)$\n",
    "<br>$\\mu$ = drift rate\n",
    "<br>$\\sigma$ = volatility\n",
    "<br>$Z=${$Z_t:t∈[0,∞)$}  is standard Brownian motion and $\\mu \\in \\mathbb{R}$ and σ∈(0,∞)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4c59ab5d-911f-46cc-86cd-acd53c21574c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It took 7.44 time units to find the value 5.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy.stats import norm\n",
    "\n",
    "\n",
    "def brownian_motion_hit_time(hit, mu=1, sigma=1, dt=0.01, t_max=100):\n",
    "    \"\"\"\n",
    "    This function models a Brownian motion process that returns the time unitl a value\n",
    "    is hit using the Arbitrage Theorem.\n",
    "    dt is the time steps.\n",
    "    hit is the value to find.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Initialize the Brownian motion process\n",
    "    t = np.arange(0, t_max+dt, dt)\n",
    "    z = norm.rvs(size=len(t), scale=np.sqrt(dt))\n",
    "    cumsum = np.cumsum((mu-0.5*sigma**2)*dt + sigma*z)\n",
    "    \n",
    "    # Find the first time the process hits the value: hit\n",
    "    hit_indices = np.where(np.abs(cumsum - hit) <= np.sqrt(dt))[0]\n",
    "    if len(hit_indices) > 0:\n",
    "        hit_time = t[hit_indices[0]]\n",
    "    else:\n",
    "        hit_time = np.nan\n",
    "    \n",
    "    return hit_time\n",
    "\n",
    "time = brownian_motion_hit_time(5)\n",
    "print('It took', time, 'time units to find the value 5.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
